js/chart_settings.js

```diff
@@ -110,46 +118,65 @@ function iso_date(date) {
 function fetch_chart_data(device) {
   dataset = [];
 
-  $.getJSON('data/' + device + '/stat.json')
-  .done(function(chart_json) {
-    $.each(chart_json, function(key, val) {
-      val.date = iso_date(val.date);
-      if (val.memory === 0) {
-        delete val.memory;
-      }
-      dataset.push(val);
-    });
+  if (!firebase.apps.length) {
+    return;
+  }
 
-    init_datepickers(dataset[0].date, dataset[dataset.length - 1].date);
+  var last_element = "";
+  var first_element = "";
 
-    var start = new Date(dataset[dataset.length - 1].date);
-    start.setMonth(start.getMonth() - 2);
-    if (start < new Date(dataset[0].date)) {
-      start = dataset[0].date;
-    }
-    $('#chart-datepicker-from').val(iso_date(start));
+  g_db_ref.child(g_db_keys[g_db_keys.length - 1]).once('value').then(function(snapshot) {
+    first_element = snapshot.val();
+    g_db_ref.child(g_db_keys[0]).once('value').then(function(snapshot) {
+      last_element = snapshot.val();
+
+      init_datepickers(iso_date(first_element.date), iso_date(last_element.date));
+
+      var start_date = new Date(last_element.date);
+      start_date.setMonth(start_date.getMonth() - 2);
+      if (start_date < new Date(first_element.date)) {
+        start_date = new Date(first_element.date);
+      }
+      else {
+        start_date = iso_date(start_date)
+      }
 
-    update_chart(start, dataset[dataset.length - 1].date);
-  })
-  .fail(function(jqxhr, textStatus, error) {
-    var err = textStatus + ', ' + error;
-    console.error('Request Failed: ' + err);
+      update_chart(start_date, last_element.date);
+
+      $('#chart-datepicker-from').val(start_date);
+    }, function(error) {
+      console.error(error);
+    });
+  }, function(error) {
+    console.error(error);
   });
 }
 
 function update_chart(from, to) {
-  var slice = [],
-      d_from = new Date(from),
-      d_to = new Date(to);
-  $.each(dataset, function(key, val) {
-    var date = new Date(val.date);
-    if (date.valueOf() >= d_from.valueOf()
-        && date.valueOf() <= d_to.valueOf())
-    {
-      slice.push(val);
-    }
-  });
 
-  generate_chart(slice, 'binary');
-  generate_chart(slice, 'memory');
+  var slice = [];
+  var d_to = new Date(to);
+  d_to.setDate(d_to.getDate() + 1);
+  g_db_ref.orderByChild('date').startAt(iso_date(from)).endAt(iso_date(d_to)).once("value", function(testcases) {
+    testcases.forEach(function (testcase) {
+      var data = testcase.val();
+      data.date = iso_date(data.date);
+
+      var tests = data.tests;
+      var total_memory = 0;
+      tests.forEach(function(testname) {
+        if (testname.memory){
+          if (!isNaN(testname.memory)) {
+            total_memory += parseInt(testname.memory)
+          }
```

Comment:
`parseInt` can also return `NaN`, you could use it to simplify the whole `forEach` logic.